sudoku_solver.cc

```diff
@@ -0,0 +1,63 @@
+#include "sudoku_solver.h"
+
+std::optional<SudokuBoard> SudokuSolver::solve_board(SudokuBoard &board) {
+  solve_board_with_max_guesses(board, board.get_missing_cells_amount());
+  if (board.calc_status() == WIN) {
+    return board;
+  }
+  return {};
+}
+
+void SudokuSolver::solve_board_with_max_guesses(SudokuBoard &board,
+                                                int max_guess_amount) {
+  if (board.calc_status() != PROG) {
+    return;
+  }
+  // TODO: possible to add here logical steps
+  make_guess(board, max_guess_amount);
+}
+
+void SudokuSolver::make_guess(SudokuBoard &board, int max_guess_amount) {
+  SudokuBoard board_without_guesses = board;
+  auto [guess_cell_r, guess_cell_c] = get_cell_indices_for_guessing(board);
+  const std::set<int> &possible_options =
+      board_without_guesses.get_cell_options(guess_cell_r, guess_cell_c);
+
+  for (int allowed_guess_amount = 1; allowed_guess_amount < max_guess_amount;
+       allowed_guess_amount++) {
+
+    auto option = possible_options.begin();
+    while (option != possible_options.end()) {
+      board.set_cell_value(guess_cell_r, guess_cell_c, *option);
+      solve_board_with_max_guesses(board, allowed_guess_amount - 1);
+      if (board.calc_status() == WIN) {
+        return;
+      }
+      // TODO: I do alot of copying in here, is there a more officient way?
```

Comment:
Yes, you can implement a more efficient `undo` move, for that you only need to keep track of the moves you did inside guesses, and not create an entire board just for a single move :)